src/lib/logging.ml

```diff
@@ -1,41 +1,38 @@
 open Ketrew_pure.Internal_pervasives
 open Unix_io
 
+(* A parametric circular buffer. *)
 module Ring = struct
   type 'a t = {
     recent: 'a option array;
     mutable index: int;
-
   }
+
   let create ?(size = 4200) () =
     {recent = Array.make size None; index = 0}
 
   let add t v =
     t.recent.(t.index) <- Some v;
-    begin match (t.index + 1) with
-    | i when i = (Array.length t.recent) -> t.index <- 0
-    | other -> t.index <- (t.index + 1);
-    end;
-    ()
+    t.index <- (t.index + 1) mod (Array.length t.recent)
 
-  let fold t ~init ~f =
+  let fold_right t ~init ~f =
+    let l = Array.length t.recent in
+    let p i = (i - 1 + l) mod l in
     let rec go count prev index =
       let next item = f prev item in
       match index, t.recent.(index) with
       | _, None -> prev
-      | _, _ when count = Array.length t.recent -> prev
-      | 0, Some s ->
-        go (count + 1) (next s) (Array.length t.recent - 1)
+      | _, _ when count = l -> prev
       | n, Some s ->
-        go (count + 1) (next s) (n - 1)
+        go (count + 1) (next s) (p n)
     in
-    if t.index = 0 then go 0 init (Array.length t.recent - 1) else go 0 init (t.index - 1) 
+    go 0 init (p t.index)
```

Comment:
I had some grand plans of replacing the `fold_right` with an actual `fold_left` since the one use case is in the logs, and there you actually do want to just traverse the messages in the correct order, as opposed to building up the list. But premature optimization and all that.